django/cantusdb_project/main_app/permissions.py

```diff
@@ -0,0 +1,139 @@
+from django.db.models import Q
+from typing import Optional
+from main_app.models import (
+    Source,
+    Chant,
+    Sequence,
+)
+from users.models import User
+
+
+def user_can_edit_chants_in_source(user: User, source: Optional[Source]) -> bool:
+    """
+    Checks if the user can edit Chants in a given Source.
+    Used in ChantDetail, ChantList, ChantCreate, ChantDelete, ChantEdit,
+    ChantEditSyllabification, and SourceDetail views.
+    """
+    if user.is_anonymous or (source is None):
+        return False
+
+    source_id = source.id
+    user_is_assigned_to_source: bool = user.sources_user_can_edit.filter(
+        id=source_id
+    ).exists()
+
+    user_is_project_manager: bool = user.groups.filter(name="project manager").exists()
+    user_is_editor: bool = user.groups.filter(name="editor").exists()
+    user_is_contributor: bool = user.groups.filter(name="contributor").exists()
+
+    return (
+        (user_is_project_manager)
+        or (user_is_editor and user_is_assigned_to_source)
+        or (user_is_editor and source.created_by == user)
+        or (user_is_contributor and user_is_assigned_to_source)
+        or (user_is_contributor and source.created_by == user)
+    )
+
+
+def user_can_proofread_chants_in_source(user: User, source: Source) -> bool:
+    """
+    Checks if the user can access the proofreading page of a given Source.
+    Used in ChantProofreadView.
+    """
+    source_id = source.id
+    if user.is_anonymous:
+        return False
+    user_is_assigned_to_source: bool = user.sources_user_can_edit.filter(
+        id=source_id
+    ).exists()
+
+    user_is_project_manager: bool = user.groups.filter(name="project manager").exists()
+    user_is_editor: bool = user.groups.filter(name="editor").exists()
+
+    return (user_is_project_manager) or (user_is_editor and user_is_assigned_to_source)
+
+
+def user_can_view_unpublished_source(user: User, source: Source) -> bool:
```

Comment:
I don't understand this function name - I'd expect `user_can_view_source`. In the last line, we `return (source.published) or (user_is_authenticated)`. We shouldn't need to check that `source.published == True`: since the function name is `user_can_view_unpublished_source`, we already know the source is unpublished.

Based on both the function name and the docstring, anyone who passes a published source into the function is not using the function properly. The function, as it currently exists, clearly accepts only unpublished sources.

We can apply similar reasoning to the two other `user_can_view_unpublished_...` below.

Reply:
Makes, sense. Based on the context of how they are used in the view, we should name it instead to `user_can_view_....` because the function should accept both published and unpublished sources.